src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs

```diff
@@ -17,34 +19,107 @@
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
+    using BoxedMemberNames = StrongBox<ImmutableSegmentedHashSet<string>>;
+
     internal sealed class DeclarationTreeBuilder : CSharpSyntaxVisitor<SingleNamespaceOrTypeDeclaration>
     {
+        /// <summary>
+        /// Cache of node to the last set of member names computed for it.  Helpful for avoiding excess hashset
+        /// allocations for common scenarios where compilations are created with the same trees used to create prior
+        /// compilations.
+        /// </summary>
+        /// <remarks>
+        /// We store as green nodes for two purposes.  First, it allows the data to be stored, without holding onto the
+        /// whole tree.  It allows files with multiple types in them to reuse the member names for untouched types when
+        /// others in the file are edited.
+        /// </remarks>
+        private static readonly ConditionalWeakTable<GreenNode, BoxedMemberNames> s_nodeToMemberNames
+            = new ConditionalWeakTable<GreenNode, BoxedMemberNames>();
+
+        private static readonly BoxedMemberNames s_emptyMemberNames = new BoxedMemberNames(ImmutableSegmentedHashSet<string>.Empty);
+
         private readonly SyntaxTree _syntaxTree;
         private readonly string _scriptClassName;
         private readonly bool _isSubmission;
 
+        /// <summary>
+        /// Stored in lexical order for the types in this tree that had member names the last time we created decls for
+        /// the tree.
+        /// </summary>
+        private readonly OneOrMany<WeakReference<BoxedMemberNames>> _previousMemberNames;
```

Comment:
Is the point here to optimize for the common case that a file contains just one type declaration?

Reply:
Yes.  No need to allocate arrays here in the common case.